### ISSUE TITLE
native crash in RealmResults#deleteAllFromRealm()

### ISSUE THREAD

**Issue by zaki50** [realm/realm-java#2759]
#### Goal

allow to call `RealmResults.deleteAllFromRealm()` against the `RealmResults` object which contains deleted object in it.

#### Expected Results

no native crash and all existing objects are deleted.

#### Actual Results

native crash

#### Steps & Code to Reproduce

executing following sample code

#### Code Sample

```java
        realm.beginTransaction();
        realm.createObject(Dog.class);
        realm.createObject(Dog.class);
        realm.commitTransaction();

        RealmResults<Dog> results = realm.where(Dog.class).findAll();

        realm.beginTransaction();

        Dog dog = results.get(0);
        dog.deleteFromRealm();

        results.deleteAllFromRealm();

        realm.commitTransaction();
```

#### Version of Realm and tooling
Realm version(s): current master

Android Studio version: 2.1

Which Android version and device: 4.1.1 on genymotion

**Comment by beeender** [comment]
i doubt we need call `sync_if_needed` in `deleteAllFromRealm()`

**Comment by zaki50** [comment]
I don't think we should call  `sync_if_needed` since it may cause unintentional deletes of objects that are not included in the results on current version, but included in results on latest version.

**Comment by cmelchior** [comment]
@zaki50 Is there core issue created for this?

**Comment by zaki50** [comment]
@cmelchior not yet

**Comment by Zhuinden** [comment]
Wait, isn't this because you're deleting invalid objects through `deleteAllFromRealm()`?

**Comment by kneth** [comment]
@zaki50 Have you been able to write a core unit test which reproduces it?

**Comment by kneth** [comment]
I have create a simple test in https://github.com/realm/realm-java/tree/kg/issue2759

**Comment by bmunkholm** [comment]
@zaki50 For these kind of issues the best is to create a PR with a unit tests that fails.

**Comment by kneth** [comment]
It seems that `sync_if_needed()` isn't called.

**Comment by Zhuinden** [comment]
@kneth As I asked, isn't it because the RealmResults is a `TableView` that is desynchronized due to not calling `syncIfNeeded()` and no longer being "live" even in a transaction, and therefore having invalid objects in the RealmResult and the native pointer being wrong due to being a previous version of the Table itself?

**Comment by cmelchior** [comment]
We have uncovered a bug in the underlying database and are working on a fix: https://github.com/realm/realm-core/issues/827

**Comment by Zhuinden** [comment]
@cmelchior  oh. Cool! Sorry about rambling then.

**Comment by zaki50** [comment]
@bmunkholm sure. I'll do in next time.

**Comment by colintheshots** [comment]
Any short term workarounds? Is there an older version which doesn't exhibit this behavior? We just spotted this bug and it's holding up a release.

**Comment by cmelchior** [comment]
It only effect you if you first delete one item, then delete all of them. If that is happening it sounds like some application logic could be rewritten instead? That said anything before 0.89 should not suffer from this.

**Comment by zaki50** [comment]
@colintheshots Current workaround is iterating results and call `deleteFromRealm()` to each  valid(`obj.isValid() == true`) object.

**Comment by beeender** [comment]
@Zhuinden 

> Now I truly do wonder if size() works on a RealmResults in a transaction after a delete.

size() return the same value but i think it is the design.

It is supposed to use `results.get(0).isValid()` to check in such case.

**Comment by kneth** [comment]
The Realm Core update merged in #2803 fixes this issue. I have added a test which is identical to the code snippet mentioned when the issue was opened.
